### PR TITLE
Add explanation and fix typo in some tutorials

### PR DESCRIPTION
**Pull request by JinraeKim**
I've followed some tutorials, and especially in comparison between manual QP differentiation and using DiffOpt,
I thought it would be better to add some explanation why `0.0 * index(x[1])` is used and what `-1.0` stands for `cause it's confusing considering the value of `h = -1.0`.

### CODE REVIEW

**Comment by JinraeKim** [comment]
I'm not familiar with generating documentation so I modified some files, which look like some source files to generate the corresponding documentation.

**Comment by odow** [comment]
> I modified some files

You chose wisely.

You can build the docs by running:
```Julia
cd ~/.julia/dev/DiffOpt
julia --project=docs docs/make.jl
```
If it complains about not being able to find DiffOpt, then
```Julia
cd ~/.julia/dev/DiffOpt
julia --project=docs

julia> ] dev .
julia> exit()

julia --project=docs docs/make.jl
```

**Comment by JinraeKim** [comment]
Should I build new documentation and add it to this PR?

As I'm not familiar with generating docs, I'm worried about it :)

**Comment by odow** [comment]
> Should I build new documentation and add it to this PR?

Nope. There is an automatic CI job that builds the documentation and publishes a new version of the website:
https://github.com/jump-dev/DiffOpt.jl/runs/4465766227?check_suite_focus=true

**Comment by JinraeKim** [comment]
@odow If you don't mind, can you answer one the following question?
Then, where can I find and check the new version of automatically generated website?

**Comment by odow** [comment]
Ah. Good question. If you had commit rights to DiffOpt, then you could view it at 
https://github.com/jump-dev/DiffOpt.jl/previews/PR177. 

But since you're an external contributor you don't have permission to publish a new version:
image

Instead, we just make sure that the CI job succeeds.

When this PR is merged it will rebuild the docs.

You don't need to do anything else for this PR. But I'll let someone who works on DiffOpt.jl review it in more detail and then merge.

**Comment by JinraeKim** [comment]
@odow 
I just wondered if I can check the auto-generated docs.
Thanks!

**Comment by odow** [comment]
Yeah for external contributors you have to build them locally to check. 

But also, it's not the end of the world if there is a small mistake that slips through the PR. You can always wait for the docs to build on `master` and then fix any mistakes in another PR. In general, I try to encourage (without much success) people to make documentation pull requests. If you just touch files in `/docs` you can't break the code so nothing bad can happen if you make a mistake.

File: docs/src/examples/matrix-inversion-manual.jl
Comment (joaquimg) [inline docs/src/examples/matrix-inversion-manual.jl:146]: Not exactly.
This is actually a perturbation on the problem. That is the direction that the derivative will be computed. In this case the direction is only following the "h[1]" axis and nothing else. There is one axis for each parameter of the problem "h[1]", "G[1,1]", "G[1,2]" and so on.

Then, the get ForwardOutVariablePrimal will return the derivative of 'x' in that direction

**Review by joaquimg — CHANGES_REQUESTED** [review]
Thanks a lot!

There is on imprecision that we should fix before merging.

File: docs/src/examples/matrix-inversion-manual.jl
Comment (JinraeKim) [reply docs/src/examples/matrix-inversion-manual.jl:146]: Thank you for your comment.
I'm really sorry that I don't get it clearly.

Can you explain why the argument should be `0.0 * index(x[1]) - 1.0` and how it changes if different parameter is taken for calculation of derivatives?

Comment (JinraeKim) [reply docs/src/examples/matrix-inversion-manual.jl:146]: Apart from that, I think that the output of `get ForwardOutVariablePrimal` returns `dx/dh` in the tutorial, right?

Comment (joaquimg) [reply docs/src/examples/matrix-inversion-manual.jl:146]: Derivatives in Rn require a direction, see:
https://en.wikipedia.org/wiki/Directional_derivative

that argument is defining that direction to be (1 * h[1], 0*G[1,1], 0*G[1,2], ...).

In other (very different) words, that carries the information of which vector is going to be multiplied by the Jacobian: (1 * dh[1], 0*dG[1,1], 0*dG[1,2], ...)

Comment (joaquimg) [reply docs/src/examples/matrix-inversion-manual.jl:146]: see also: https://math.stackexchange.com/questions/3191003/directional-derivative-and-jacobian-matrix

File: docs/src/examples/custom-relu.jl
Comment (matbesancon) [inline docs/src/examples/custom-relu.jl:138]:
```suggestion
# the number of epochs `epoch` and the connectivity `inner`.
```

File: docs/src/examples/matrix-inversion-manual.jl
Comment (JinraeKim) [reply docs/src/examples/matrix-inversion-manual.jl:146]: Thank you, @joaquimg.
But my question is, in that line, "how to let the code know the direction" for directional derivative calculation.
That is,
I don't understand how the line `0.0 * index(x[1]) - 1.0` indicates that "the direction is (1, 0, 0, ...) with axes of (h[1], G[1, 1], G[1, 2], ...)".

Comment (JinraeKim) [reply docs/src/examples/matrix-inversion-manual.jl:146]: What if I want to decide the direction as (0, 1, 0) with axes of (h[1], G[1, 1], G[1, 2])?

Comment (joaquimg) [reply docs/src/examples/matrix-inversion-manual.jl:146]: In that case you would do:
1.0x[1] + 0.0

cons[1] means that you are selecting the first constraint and x[1] means the coefficient of the first variable: that is G[1,1].

The 1.0 multiplying x[1] means 1 in that direcation.

Comment (JinraeKim) [reply docs/src/examples/matrix-inversion-manual.jl:146]: Well, I see.
Then, you mean that when considering the constraint `Gx - h <=0`, the direction of `G[1, i]` corresponds to `x[i]` and `h` corresponds to the constant...?
Also,
1) the `-1.0` in `0.0 * index(x[1]) - 1.0` corresponds to "1 in h-direction", or "-h"-direction? (it should be "h-direction" in context, probably due to the negative sign convention of `h` in `Gx - h <= 0`.
2) `0.0 * index(x[1])` is still to make the type a AbstractScalarSomething? (I forgot the name, sorry)

Comment (JinraeKim) [reply docs/src/examples/matrix-inversion-manual.jl:146]: EDIT:
I'm not sure but from my observation that the `0.0 * index(x[1]) - 1.0` corresponds to `dG * x - dh`; probably that's why @joaquimg referred to as "directional derivative".
This can be found in some source codes like [how to get `dh`](https://github.com/jump-dev/DiffOpt.jl/blob/cf808b04c122a73500cdd4db9bc269b63ee3b709/src/diff_opt.jl#L689).

Please inform me if I misunderstood.

Comment (joaquimg) [reply docs/src/examples/matrix-inversion-manual.jl:146]: the sign on the constant is a bit non-trivial, see my own reflections here: https://github.com/jump-dev/DiffOpt.jl/issues/174
might help...

Comment (JinraeKim) [reply docs/src/examples/matrix-inversion-manual.jl:146]: @joaquimg #174 provides a very good insight of how it works.
It would be better to read it before reading the source codes... 👍

Comment (JinraeKim) [reply docs/src/examples/matrix-inversion-manual.jl:146]: @joaquimg 
Can you check whether I understood it correctly?

The constraint function is basically the form of `f(x, G, h_) = Gx + h_` internally,
so to get the directional derivative `h`-axis (with unit vector), we have to flip the sign of the constant as we consider the constraint of `f(x, G, -h) = Gx - h`.
Also, to get directional derivatives of other axes, in the case of `G[1, i]`, just passing `x[i]` into where `0.0 * index(x[1]) - 1.0` is placed provides the corresponding directional derivative (with unit vector).
Multiplication by scalar replaces the unit vector condition intuitively.

Comment (joaquimg) [reply docs/src/examples/matrix-inversion-manual.jl:146]: looks good!

Comment (JinraeKim) [inline docs/src/examples/matrix-inversion-manual.jl:146]:
```suggestion
    0.0 * index(x[1]) - 1.0,  # to indicate the direction vector to get directional derivatives
```

File: docs/src/examples/matrix-inversion-manual.jl
Comment (JinraeKim) [inline docs/src/examples/matrix-inversion-manual.jl:149]:
```suggestion
# Note that `0.0 * index(x[1])` is used to make its type `typeof(0.0 * index(x[1]) - 1.0) <: MOI.AbstractScalarFunction`.
# To indicate different direction to get directional derivative, users should replace `0.0 * index(x[1]) - 1.0` as the form of `A*x - b`, where `A` and `b` correspond to the elements of direction vectors along `G` and `h` axes, respectively.
```

**Review by JinraeKim — COMMENTED** [review]
@joaquimg
I modified the comments related to directional derivatives.

File: docs/src/examples/matrix-inversion-manual.jl
Comment (joaquimg) [reply docs/src/examples/matrix-inversion-manual.jl:146]: looks good for me

File: docs/src/examples/matrix-inversion-manual.jl
Comment (joaquimg) [reply docs/src/examples/matrix-inversion-manual.jl:149]: I would avoid using A and b, since they are used to describe equality constraints.
maybe `dG * x - dh`?

Comment (JinraeKim) [reply docs/src/examples/matrix-inversion-manual.jl:149]: IMO, it may also be confusing as in #174 we will finally get `dx/dy` where `y` denotes perturbation itself.
That is, if we use the notation as `dG` and `dh`, we will get something like `dx/ddh`.

However, I agree that still `dG` and `dh` are intuitive and convenient.

Comment (joaquimg) [reply docs/src/examples/matrix-inversion-manual.jl:149]: Forward mode is usually simpler to understand. dG and dh also follow (closely) the notation of the paper OptNet

Comment (JinraeKim) [reply docs/src/examples/matrix-inversion-manual.jl:149]: Your comment is reflected.

File: docs/src/examples/matrix-inversion-manual.jl
Comment (JinraeKim) [reply docs/src/examples/matrix-inversion-manual.jl]: For the change request, see the following commits.

**Review by joaquimg — APPROVED** [review]

**Comment by matbesancon** [comment]
Thanks for the PR!

**Pull request merged**
